File: src/metabolism.cc

#include "metabolism.hh"

std::tuple<vector_t, vector_t, vector_t>
initialize 
    ( const vector_t& variable_metabolites
    , const vector_t& fixed_metabolites_log_counts
    , const vector_t& target_log_counts
    , const matrix_t& stoichiometric_matrix
    , const vector_t& equilibrium_constants
    )
{
    //
    // modify me for better initialization
    //
    auto initial_values = find_initial_values
        ( variable_metabolites
        , target_log_counts
        , fixed_metabolites_log_counts
        , stoichiometric_matrix
        , equilibrium_constants
        );

    vector_t beta_init = std::get<0>(initial_values);
    vector_t flux_init = std::get<1>(initial_values);
    vector_t variable_metabolites_init = std::get<2>(initial_values);
    return { beta_init, flux_init, variable_metabolites_init };
}

std::tuple<vector_t, vector_t, vector_t, vector_t>
optimize
    ( const vector_t& variable_metabolites_init
    , const vector_t& flux_variables_init
    , const vector_t& beta_init
    , const vector_t& fixed_metabolites_log_counts
    , const vector_t& target_log_variable_metabolites_counts
    , const matrix_t& stoichiometric_matrix
    , const vector_t& equilibrium_constants
    , const index_list_t& objective_reaction_indices
    )
{
    auto results_tuple = maximum_entropy_solver 
        ( variable_metabolites_init
        , fixed_metabolites_log_counts
        , flux_variables_init
        , beta_init
        , target_log_variable_metabolites_counts
        , stoichiometric_matrix
        , equilibrium_constants
        , objective_reaction_indices
        );
    vector_t steady_state_sol = std::get<0>(results_tuple);
    vector_t flux_sol = std::get<1>(results_tuple);
    vector_t alpha_sol = std::get<2>(results_tuple);
    vector_t h_sol = std::get<3>(results_tuple);
    vector_t beta_sol = std::get<4>(results_tuple);
    vector_t variable_metabolites_sol = std::get<5>(results_tuple);

    vector_t e_regulation = alpha_sol;
    matrix_t positive_stoich_matrix = stoichiometric_matrix;
    for ( auto row : positive_stoich_matrix.rowwise()) {
        for (auto elem : row) {
            if (elem < 0.0) {
                elem = 0.0;
            }
        }
    }
    matrix_t negative_stoich_matrix = stoichiometric_matrix;
    for ( auto row : negative_stoich_matrix.rowwise()) {
        for (auto elem : row) {
            if (elem > 0.0) {
                elem = 0.0;
            }
        }
    }
    int num_total_metabolites = variable_metabolites_sol.size() + fixed_metabolites_log_counts.size();
    vector_t all_metabolites(num_total_metabolites);
    all_metabolites << variable_metabolites_sol, fixed_metabolites_log_counts;

    /*
    auto delta = vector_t::Zero(num_total_metabolites);
    double mu0 = 1.0;
    rxn_flux = odds_diff
        ( variable_metabolites_sol
        , fixed_metabolites_log_counts
        , mu0
        , stoichiometric_matrix
        , negative_stoich_matrix
        , positive_stoich_matrix
        , delta
        , equilibrium_constants
        , e_regulation
        );
    */

    return { steady_state_sol, flux_sol, variable_metabolites_sol, alpha_sol};
}

vector_t
odds_diff
    ( const vector_t& variable_metabolites_counts
    , const vector_t& fixed_metabolites_log_counts
    , const double mu0
    , const matrix_t& stoichiometric_matrix
    , const matrix_t& negative_stoich_matrix
    , const matrix_t& positive_stoich_matrix
    , const vector_t& delta
    , const vector_t& equilibrium_constants
    , const vector_t& e_regulation
    )
{
    int num_total_metabolites = variable_metabolites_counts.size() + fixed_metabolites_log_counts.size();
    vector_t all_metabolites(num_total_metabolites);
    all_metabolites << variable_metabolites_counts, fixed_metabolites_log_counts;
    double direction = 1.0;
    vector_t KQ_f = calc_odds
        ( all_metabolites
        , negative_stoich_matrix
        , positive_stoich_matrix
        , delta
        , equilibrium_constants
        , direction
        );
    vector_t equilibrium_constants_inverse = equilibrium_constants.array().inverse().matrix();
    direction = -1.0;
    vector_t KQ_r = calc_odds
        ( all_metabolites
        , negative_stoich_matrix
        , positive_stoich_matrix
        , delta
        , equilibrium_constants_inverse
        , direction
        );
    vector_t KQ_diff = ( e_regulation.array() * (KQ_f - KQ_r).array() ).matrix();
    
    return KQ_diff;
}

vector_t
calc_odds
    ( const vector_t& log_counts
    , const matrix_t& negative_stoich_matrix
    , const matrix_t& positive_stoich_matrix
    , const vector_t& delta
    , const vector_t& equilibrium_constants
    , const double direction
    )
{
    vector_t counts = log_counts.array().exp().matrix();
    vector_t delta_counts = counts + delta;
    vector_t log_delta = delta_counts.array().log().matrix();
    vector_t Q_inv = 
        ( -direction * (negative_stoich_matrix * log_counts)
        + (positive_stoich_matrix * log_delta)
        ).array().exp().matrix();
    vector_t KQ = (equilibrium_constants.array() * Q_inv.array()).matrix();
    return KQ;
}

vector_t
scale_flux
    ( const vector_t& reaction_flux
    , const int iglucose
    , const double Vmax
    , const double Km
    , const double s
    )
{
    double V = Vmax * s/(Km + s);
    vector_t scaled_flux = V/reaction_flux(iglucose) * reaction_flux;

    return reaction_flux;
}

std::tuple<vector_t, vector_t>
run_metabolism
    ( const vector_t& variable_metabolites
    , const vector_t& fixed_metabolites_log_counts
    , const vector_t& target_log_variable_metabolites_count
    , const matrix_t& stoichiometric_matrix
    , const vector_t& equilibrium_constants
    , const index_list_t& objective_reaction_indices
    )
{
    // get initial values
    auto initial_results = initialize
        ( variable_metabolites
        , fixed_metabolites_log_counts
        , target_log_variable_metabolites_count
        , stoichiometric_matrix
        , equilibrium_constants
        );
    vector_t beta_init = std::get<0>(initial_results);
    vector_t flux_values_init = std::get<1>(initial_results);
    vector_t variable_metabolites_init = std::get<2>(initial_results);

    // get solution
    auto optimize_results = optimize
        ( variable_metabolites_init
        , flux_values_init
        , beta_init
        , fixed_metabolites_log_counts
        , target_log_variable_metabolites_count
        , stoichiometric_matrix
        , equilibrium_constants
        , objective_reaction_indices
        );
    vector_t steady_state_sol = std::get<0>(optimize_results);
    vector_t flux_sol = std::get<1>(optimize_results);
    vector_t variable_metabolites_sol = std::get<2>(optimize_results);
    vector_t e_regulation = std::get<3>(optimize_results);


    // run odds difference
    matrix_t positive_stoich_matrix = stoichiometric_matrix;
    for ( auto row : positive_stoich_matrix.rowwise()) {
        for (auto elem : row) {
            if (elem < 0.0) {
                elem = 0.0;
            }
        }
    }
    matrix_t negative_stoich_matrix = stoichiometric_matrix;
    for ( auto row : negative_stoich_matrix.rowwise()) {
        for (auto elem : row) {
            if (elem > 0.0) {
                elem = 0.0;
            }
        }
    }
    int num_total_metabolites = variable_metabolites.size() + fixed_metabolites_log_counts.size();
    auto delta = vector_t::Zero(num_total_metabolites);
    double mu0 = 1.0;
    vector_t reaction_flux = odds_diff
        ( variable_metabolites
        , fixed_metabolites_log_counts
        , mu0
        , stoichiometric_matrix
        , negative_stoich_matrix
        , positive_stoich_matrix
        , delta
        , equilibrium_constants
        , e_regulation
        );

    return { reaction_flux, variable_metabolites_sol };
}

File: python/metabolism.py
import numpy as np
import pyomo
import pyomo.environ as pe
import itertools
import numpy.random as nprd
from scipy.linalg import norm
from scipy.optimize import least_squares
import scipy.linalg as spL
import pyutilib.services
from pyomo.opt import TerminationCondition
from importlib import reload
import timeit

import maximum_entropy_pyomo_relaxed as mep
import feasible_point_pyomo_relaxed as fep
mep = reload(mep)


def initialize(v_log_counts,f_log_counts,target_log_vcounts, S, K):

  ### SOLVE FOR AN INITIAL FEASIBLE POINT
  [solved,feas_obj,bi_sol, yi_sol, alphai_sol, hi_sol, betai_sol, ni_sol] = fep.Feasible_point_solver(v_log_counts, target_log_vcounts, f_log_counts, S, K)                            

  v_log_counts = ni_sol

  # find out if it terminated on optimal solution
  print(feas_obj)
  if solved == 0:
    print("solver failed to converge")
  else: 
    print("Solver Converged")

  #set an objective tolerance to call solution feasible
  feasible_tol = 1e-12

  if feas_obj > feasible_tol:
    print("Solver did not find a feasible point")
    print("Feasibility tolerance = ", feasible_tol, "\tObjective tolerance = ", feas_obj)    

  # set the initial condition as the solution from the feasible point solve
  n_ini = np.reshape(ni_sol,(len(ni_sol),) )
  b_ini = np.reshape(bi_sol,(len(bi_sol),) )
  y_ini = np.reshape(yi_sol,(len(yi_sol),) )
  beta_ini = np.reshape(betai_sol,(len(betai_sol),) )
  h_ini = np.reshape(hi_sol,(len(hi_sol),) )

  # save for c++
  path = "../data/python_feasible_point/"
  np.savetxt(path + "variable_metabolites_log_counts.csv",n_ini, delimiter=',')
  np.savetxt(path + "flux_variables.csv", y_ini, delimiter=',')
  np.savetxt(path + "null_space_variables.csv",beta_ini, delimiter=',')

  return(beta_ini, y_ini,n_ini)

def optimize(n_init, y_init, beta_init, f_log_counts,target_log_vcounts, S, K,obj_rxn_idx, max_iter = 10000):

  start_t = timeit.timeit()
  [b_sol, y_sol, alpha_sol, h_sol, beta_sol, n_sol] = mep.Max_ent_solver(n_init,y_init,beta_init,target_log_vcounts, f_log_counts, S, K, obj_rxn_idx)
  end_t = timeit.timeit()
  print('time Max_ent_solver =', end_t - start_t)

  E_regulation = alpha_sol
  v_log_counts = n_sol
  P = np.where(S>0,S,0)
  R = np.where(S<0, S, 0)
  log_counts = x = np.concatenate((v_log_counts, f_log_counts))
  delta = delta = np.zeros(log_counts.size)
  rxn_flux = oddsDiff(v_log_counts, f_log_counts, 1, S, R, P, delta, K, E_regulation)
  dndt = S.T.dot(rxn_flux)
  #print('Rxn Flux:', rxn_flux)
  #print('Metabolite Derivatives:',dndt)

  return (b_sol, y_sol, n_sol, alpha_sol)

def oddsDiff(vcounts,fcounts,mu0,S, R, P, delta,Keq,E_Regulation):
  metabolites = np.append(vcounts,fcounts)
  KQ_f = odds(metabolites,mu0,S, R, P, delta,Keq);
  Keq_inverse = np.power(Keq,-1);
  KQ_r = odds(metabolites,mu0,-S, P, R, delta,Keq_inverse,-1);
  
  #WARNING: Multiply regulation here, not on individual Keq values.
  KQdiff =  E_Regulation * (KQ_f - KQ_r);
  return(KQdiff)

def odds(log_counts,mu0,S, R, P, delta, K, direction = 1):
  counts = np.exp(log_counts)
  delta_counts = counts+delta;
  log_delta = np.log(delta_counts);
  Q_inv = np.exp(-direction*(R.dot(log_counts) + P.dot(log_delta)))
  KQ = np.multiply(K,Q_inv);
  return(KQ)

def scale_flux(rxn_flux,iglucose, Vmax, Km, s):
#  inputs:
#  Kmichealis constant, Km
#  Vmax, Vmax
# External glucose concentration, s
# index of glucose uptake reaction, iglucose
# reaction flux from optimization, rxn_flux

  V = Vmax * s/(Km + s)
  scaled_flux = V/rxn_flux[iglucose] * rxn_flux
  return(scaled_flux)

def run(v_log_counts,f_log_counts,target_log_vcounts, S, K, iuptake, Vmax, Km, s, obj_rxn_idx):

  beta_init, y_init,n_init = initialize(v_log_counts,f_log_counts,target_log_vcounts, S, K)
  b_sol, y_sol, n_sol, alpha_sol = optimize(n_init, y_init, beta_init, f_log_counts,target_log_vcounts, S, K,obj_rxn_idx)

  E_regulation = alpha_sol
  v_log_counts = n_sol
  P = np.where(S>0,S,0)
  R = np.where(S<0, S, 0)
  log_counts = x = np.concatenate((v_log_counts, f_log_counts))
  delta = delta = np.zeros(log_counts.size)
  rxn_flux = oddsDiff(v_log_counts, f_log_counts, 1, S, R, P, delta, K, E_regulation)
  #dndt = S.T.dot(rxn_flux)

  #scaled_flux = scale_flux(rxn_flux,iuptake, Vmax, Km, s)
  return(rxn_flux,n_sol)

  

#def report():
#  inputs:
#  Kmichealis
#  Vmax
#  index of cell wall
#  index of glucose
#  alpha_sol - regulation
#  n_sol - variable log counts of metabolites 
#  f_log_counts - fixed log counts of metabolites
#  K, S - equilibrium constants and stoichiometric matrix

#  compute P, R, delta
#  E_regulation_pyomo = alpha_sol
#  rxn_flux_ip = oddsDiff(n_sol, f_log_counts, mu0, S, R, P, delta, K, E_regulation_pyomo)
#  print('Metabolite Derivatives:')
#   V = V_max*s/(km + s)
#  dndt = S_active.T.dot(rxn_flux_ip)
#  print(dndt)

  #output:
#  return(V, glucose_uptake, cell_wall_output)
